pkg/cloud/aliyunprovider.go

```diff
@@ -0,0 +1,828 @@
+		OSType:          node.OSType,
+	}
+}
+
+// AliyunDiskAttributes represents metadata about the product used to map to a PV.
+// Basic Attributes needed atleast to get the keys, Some attributes from k8s Node response
+// be populated directly into *PV object.
+type AliyunPVAttributes struct {
+	DiskType         int32  `json:"diskType"`
+	DiskCategory     string `json:"diskCategory"`
+	PerformanceLevel string `json:"performanceLevel"`
+}
+
+// Stage 1 support will be Pay-As-You-Go with HourlyPrice equal to TradePrice with PriceUnit as Hour
+// TO-DO: Subscription and Premptible support, need to find how to distinguish node into these categories]
+// TO-DO: Open question Subscription would be either Monthly or Yearly, Firstly Data retrieval/population
+// TO-DO:  need to be tested from describe price API, but how would you calculate hourly price, is it PRICE_YEARLY/HOURS_IN_THE_YEAR?
+type AliyunPricingDetails struct {
+	HourlyPrice  float32 `json:"hourlyPrice"`  // Represents hourly price for the given Aliyun Product
+	PriceUnit    string  `json:"priceUnit"`    // Represents the unit in which Alibaba Product is billed can be Hour, Month or Year based on the billingMethod
+	TradePrice   float32 `json:"tradePrice"`   // Original Price used to acquire the Alibaba Product
+	CurrencyCode string  `json:"currencyCode"` // Represents the currency unit of the
+}
+
+func NewAliyunPricingDetails(hourlyPrice float32, priceUnit string, tradePrice float32, currencyCode string) *AliyunPricingDetails {
+	return &AliyunPricingDetails{
+		HourlyPrice:  hourlyPrice,
+		PriceUnit:    priceUnit,
+		TradePrice:   tradePrice,
+		CurrencyCode: currencyCode,
+	}
+}
+
+// AliyunPricingTerms can have three types of supported billing method Pay-As-You-Go, Subscription and Premptible
+type AliyunPricingTerms struct {
+	BillingMethod  string                `json:"billingMethod"`
+	PricingDetails *AliyunPricingDetails `json:"pricingDetails"`
+}
+
+func NewAliyunPricingTerms(billingMethod string, pricingDetails *AliyunPricingDetails) *AliyunPricingTerms {
+	return &AliyunPricingTerms{
+		BillingMethod:  billingMethod,
+		PricingDetails: pricingDetails,
+	}
+}
+
+// Alibaba Pricing struct carry the Attributes and pricing information for Node or PV
+type AliyunPricing struct {
+	NodeAttributes *AliyunNodeAttributes
+	PVAttributes   *AliyunPVAttributes
+	PricingTerms   *AliyunPricingTerms
+	Node           *Node
+	PV             *PV
+}
+
+// Aliyun's Provider struct
+type Aliyun struct {
+	Pricing                 map[string]*AliyunPricing // Data to store Aliyun(Alibaba cloud) pricing struct
+	DownloadPricingDataLock sync.RWMutex              // Lock Needed to provide thread safe
+	Clientset               clustercache.ClusterCache
+	Config                  *ProviderConfig
+	serviceAccountChecks    *ServiceAccountChecks
+	clusterAccountId        string
+	clusterRegion           string
+	loadedAccessKey         bool                             // Check if Aliyun is authenticated
```

Comment:
I like to use "is" for variables like this, so you get something like `Aliyun.isAccessKeyLoaded` or `Aliyun.accessKeyIsLoaded`. That makes it really clear that its a boolean flag. No requirement here, just a thought!

Reply:
And perhaps another question -- is this variable always equivalent to `Aliyun.accessKey != nil`? If so, maybe we don't need this variable at all and could replace it with a method like `func (a *Aliyun) accessKeyIsLoaded() { return a.accessKey != nil }`